source/renderer/app/components/widgets/forms/PinCode.js

```diff
@@ -31,42 +38,198 @@ export default class PinCode extends Component<Props> {
   inputsRef = [];
   focusKey = 0;
   add = false;
+  fromBackspace = false;
+
+  state = {
+    isBackSpace: false,
+    focusKeyChanged: false,
+  };
+
+  inputHasNewValue = (inputNewValue: string, key: number) => {
+    const { value } = this.props;
+    const valueIsEmpty =
+      !value[key] || value[key] === '' || inputNewValue === '';
+    const valueHasChanged =
+      inputNewValue &&
+      inputNewValue.length === 1 &&
+      value[key] !== inputNewValue;
+    return valueIsEmpty || valueHasChanged;
+  };
+
+  inputIsMarked = (input: Field) => {
+    const { inputElement } = input;
+    const { current } = inputElement;
+    return current.selectionStart !== current.selectionEnd;
+  };
+
+  isInputFieldDisabled = (
+    index: number,
+    value: Array<string>,
+    disabled: boolean
+  ) => {
+    return (
+      disabled ||
+      (index !== 0 && (!value || !value[index - 1])) ||
+      (index > 1 && index <= 4 && !value[index - 2]) ||
+      (index === 3 && value[index - 3] === '')
+    );
+  };
 
-  onChange = (inputValue: ?number, key: number) => {
+  onChange = (inputValue: ?number | ?string, key: number) => {
     const { value, onChange } = this.props;
+    const { isBackSpace } = this.state;
     const inputNewValue =
-      inputValue !== null && inputValue !== undefined
+      inputValue !== null && inputValue !== undefined && !isNaN(inputValue)
         ? inputValue.toString()
         : '';
-
-    if (
-      !Object.prototype.hasOwnProperty.call(value, key) ||
-      value[key] === '' ||
-      inputNewValue === ''
-    ) {
+    if (this.inputHasNewValue(inputNewValue, key)) {
       const newValue = value;
-      newValue[key] = inputNewValue;
+      if (!isNaN(inputValue)) {
```

Comment:
Please document the code below with comments … there is a lot of very low-level stuff going on which is hard to understand for a reader. Probably would also be good to refactor some parts into helper functions and give them speaking names + doc comments.